### ISSUE TITLE
touch does not check lock_version

### ISSUE THREAD

**Issue by gucki** [rails/rails#21698]
irb(main):002:0> Server.first.touch
      Server Load (1.2ms)  SELECT  "servers".* FROM "servers"  ORDER BY "servers"."id" ASC LIMIT 1
       (0.3ms)  BEGIN
      SQL (0.8ms)  UPDATE "servers" SET "updated_at" = '2015-09-21 08:35:34.408236', "lock_version" = 29 WHERE "servers"."id" = $1  [["id", "04531399-6f25-41e2-862a-0b542c8f805b"]]
       (5.5ms)  COMMIT
    => true

The `where` clause `AND lock_version = 28` is missing. Rails 4.2.4

**Comment by gucki** [comment]
Just for comparison, which shows the correct behavior:

    irb(main):001:0> Server.first.update_attributes!(:updated_at => Time.now)
      Server Load (1.3ms)  SELECT  "servers".* FROM "servers"  ORDER BY "servers"."id" ASC LIMIT 1
       (0.2ms)  BEGIN
      SQL (0.4ms)  UPDATE "servers" SET "updated_at" = '2015-09-21 08:40:25.839913', "lock_version" = 30 WHERE "servers"."id" = $1 AND "servers"."lock_version" = $2  [["id", "04531399-6f25-41e2-862a-0b542c8f805b"], ["lock_version", 29]]
       (2.6ms)  COMMIT
    => true

**Comment by meinac** [comment]
I've tested this with Rails versions `4.2.2`, `4.2.3`, `4.2.4` and `master`. `master` branch seems ok but the others have this issue.

**Comment by jaredbeck** [comment]
If it helps, here are the same queries Corin gave above, formatted for readability.

```sql
-- Correct
UPDATE "servers" 
SET "updated_at" = '2015-09-21 08:40:25.839913', 
  "lock_version" = 30 
WHERE "servers"."id" = $1 
  AND "servers"."lock_version" = $2  
;
-- [["id", "04531399-6f25-41e2-862a-0b542c8f805b"], ["lock_version", 29]]

-- Incorrect
UPDATE "servers" 
SET "updated_at" = '2015-09-21 08:35:34.408236', 
  "lock_version" = 29 
WHERE "servers"."id" = $1  
;
-- [["id", "04531399-6f25-41e2-862a-0b542c8f805b"]]
```

**Comment by sgrif** [comment]
An executable test case using [our bug report template](https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb) would be very helpful!

**Comment by sgrif** [comment]
Went ahead and created one. https://gist.github.com/sgrif/a466bfabb57c6dfa977a

**Comment by sgrif** [comment]
Unfortunately this cannot be fixed on 4.2. It will cause regressions as sometimes creating an association will touch a record multiple times. Fixing this requires a new feature for Rails 5, added in https://github.com/rails/rails/pull/19324, which cannot be backported.
